mailpoet/lib/Captcha/Validator/RecaptchaValidator.php

```diff
@@ -1,10 +1,10 @@
 <?php // phpcs:ignore SlevomatCodingStandard.TypeHints.DeclareStrictTypes.DeclareStrictTypesMissing
 
-namespace MailPoet\Subscription\Captcha\Validator;
+namespace MailPoet\Captcha\Validator;
 
 use MailPoet\Captcha\ReCaptchaValidator as Validator;
 
-class RecaptchaValidator implements CaptchaValidator {
+class RecaptchaValidator {
 
   /** @var Validator */
   private $validator;
```

Comment:
> You removed the CaptchaValidator interface here https://github.com/mailpoet/mailpoet/commit/744c59033b1e199c6df760925ead56853234154e and renamed the class here https://github.com/mailpoet/mailpoet/commit/47787ca2b6445e986594af1ad54d7628f7d0c3f2

Absolutely. 

I made the change in one commit initially, but during the rebase, I had to fix a conflict in the first commit. Not sure what went wrong, but some of the changes made it to the next commit :man_shrugging: 

> I believe keeping the CaptchaValidator is helpful for code readability and ensures a consistent class interface and experience for other captcha subclasses.

I see your point. I don't have a hard opinion here, happy to revert that :wink: 

I have a slight preference to keep the change though. Ensuring a "single consistent method name" between polymorphic classes doesn't sound like a good reason for an interface. To me, interfaces are meant for polymorphic references. Yet. If your argument aligns with similar instances elsewhere in the codebase, I'd favor consistency above all.

> Here, you also renamed BuiltInCaptchaValidator (which is more explicit about its intended purpose) with CaptchaValidator - This makes it a bit confusing now that we have just CaptchaValidator and RecaptchaValidator (which is what? What does it do? etc.).

I'd reach the same conclusion if I looked at it in isolation. Yet. Every other class related to the built-in CAPTCHA start with `Captcha`. I believe this change maintains consistency:

- CaptchaConstants.php
- CaptchaPhrase.php
- CaptchaRenderer.php
- CaptchaSession.php
- CaptchaFormRenderer.php

> Additionally, you renamed the class but left both the unit and integrated test classes the same. May I ask why that is so? It can increase some of the confusion I previously highlighted.

Oops, I missed that. I think I relied too much on the IDE's refactoring tool.

Reply:
> I have a slight preference to keep the change though. Ensuring a "single consistent method name" between polymorphic classes doesn't sound like a good reason for an interface. To me, interfaces are meant for polymorphic references. Yet. If your argument aligns with similar instances elsewhere in the codebase, I'd favor consistency above all.

I may not fully understand what you mean by `polymorphic classes` and `polymorphic references`. 
From my understanding of [Polymorphism](https://en.wikipedia.org/wiki/Polymorphism_(computer_science)), it appears we are using it correctly.

I believe it's nice to align with the rest of the use case in the codebase.

> I'd reach the same conclusion if I looked at it in isolation. Yet. Every other class related to the built-in CAPTCHA start with Captcha. I believe this change maintains consistency:
> 
> CaptchaConstants.php
CaptchaPhrase.php
CaptchaRenderer.php
CaptchaSession.php
CaptchaFormRenderer.php


Oh, this makes sense in that case. Are these classes not used by other captcha methods?


> Oops, I missed that. I think I relied too much on the IDE's refactoring tool.

Not a problem. This happens sometimes.

Reply:
Let me explain my argument with an example:

```
<?php

interface Shape {
    public function draw();
}

class Circle implements Shape {
    public function draw() {
        echo "Drawing a Circle\n";
    }
}

class Rectangle implements Shape {
    public function draw() {
        echo "Drawing a Rectangle\n";
    }
}

// Here's a polymorphic reference.
function renderShape(Shape $shape) {
    $shape->draw();
}
```

In this example, `renderShape` expect an interface argument. That's a usage.

The `CaptchaValidator` interface is not referenced anywhere. Absolutely no usage. The relevant logic is not referencing the interface, but implementation classes :shrug:

```
private function validateCaptcha($captchaSettings, $data): array {
  if (empty($captchaSettings['type'])) {
    return [];
  }

  try {
    if ($captchaSettings['type'] === CaptchaConstants::TYPE_BUILTIN) {
      $this->builtInCaptchaValidator->validate($data);
    }
    if (CaptchaConstants::isReCaptcha($captchaSettings['type'])) {
      $this->recaptchaValidator->validate($data);
    }
  } catch (ValidationError $error) {
    return $error->getMeta();
  }
  return [];
}
```

Something like this would justify the interface:

```
private function validateCaptcha(CaptchaValidator $validator): array {
  $this->validate($data);
}
```

I checked the codebase. It seems like interfaces aligns with my argument here. For example, `MailerMethod` class is referenced [here](https://github.com/mailpoet/mailpoet/blob/13bf305aeb29bbadd0695ee02a3735e62cc4f21f/mailpoet/lib/Mailer/Mailer.php#L20-L24).

This interface might have been used properly before. This no longer holds.

Reply:
Please find usage of `CaptchaValidator` here
* https://github.com/mailpoet/mailpoet/blob/324e02f2cfa0f2d775236c16b41433f904f7f92c/mailpoet/lib/Subscription/Captcha/Validator/RecaptchaValidator.php#L7
*  https://github.com/mailpoet/mailpoet/blob/324e02f2cfa0f2d775236c16b41433f904f7f92c/mailpoet/lib/Subscription/Captcha/Validator/BuiltInCaptchaValidator.php#L12


By the way, `CaptchaValidator` is not a hard stop for me. I was mostly concerned about the `BuiltInCaptchaValidator` rename, which will cause confusion when paired with other captcha options.

Reply:
That's cheating. Implementation references doesn't count, not for the purpose of this discussion :wink: